samples/nrf9160/modem_shell/src/cloud/cloud_shell.c

```diff
@@ -0,0 +1,195 @@
+	if (err) {
+		mosh_error("Error requesting notification of prediction availability: %d", err);
+	}
+}
+#endif /* defined(CONFIG_NRF_CLOUD_PGPS) */
+
+static void nrf_cloud_event_handler(const struct nrf_cloud_evt *evt)
+{
+	int err = 0;
+	const int reconnection_delay = 10;
+
+	switch (evt->type) {
+	case NRF_CLOUD_EVT_TRANSPORT_CONNECTING:
+		mosh_print("NRF_CLOUD_EVT_TRANSPORT_CONNECTING");
+		break;
+	case NRF_CLOUD_EVT_TRANSPORT_CONNECTED:
+		mosh_print("NRF_CLOUD_EVT_TRANSPORT_CONNECTED");
+		break;
+	case NRF_CLOUD_EVT_READY:
+		mosh_print("NRF_CLOUD_EVT_READY");
+		break;
+	case NRF_CLOUD_EVT_TRANSPORT_DISCONNECTED:
+		mosh_print("NRF_CLOUD_EVT_TRANSPORT_DISCONNECTED");
+		mosh_print("Reconnecting in %d seconds...", reconnection_delay);
+		k_work_reschedule(&cloud_reconnect_work, K_SECONDS(reconnection_delay));
+		break;
+	case NRF_CLOUD_EVT_ERROR:
+		mosh_print("NRF_CLOUD_EVT_ERROR");
+		break;
+	case NRF_CLOUD_EVT_SENSOR_DATA_ACK:
+		mosh_print("NRF_CLOUD_EVT_SENSOR_DATA_ACK");
+		break;
+	case NRF_CLOUD_EVT_FOTA_DONE:
+		mosh_print("NRF_CLOUD_EVT_FOTA_DONE");
+		break;
+	case NRF_CLOUD_EVT_RX_DATA:
+		mosh_print("NRF_CLOUD_EVT_RX_DATA");
+		if (((char *)evt->data.ptr)[0] == '{') {
+			/* Not A-GPS data */
+			break;
+		}
+#if defined(CONFIG_NRF_CLOUD_AGPS)
+		err = nrf_cloud_agps_process((char *)evt->data.ptr, evt->data.len);
+		if (!err) {
+			mosh_print("A-GPS data processed");
+#if defined(CONFIG_NRF_CLOUD_PGPS)
+			/* call us back when prediction is ready */
+			k_work_submit(&notify_pgps_work);
+#endif
+			/* data was valid; no need to pass to other handlers */
+			break;
+		}
+#endif
+#if defined(CONFIG_NRF_CLOUD_PGPS)
+		err = nrf_cloud_pgps_process(evt->data.ptr, evt->data.len);
+		if (err) {
+			mosh_error("Error processing P-GPS packet: %d", err);
+		}
+#else
+		if (err) {
+			mosh_print("Unable to process A-GPS data, error: %d", err);
+		}
+#endif
+		break;
+	case NRF_CLOUD_EVT_USER_ASSOCIATION_REQUEST:
+		mosh_print("NRF_CLOUD_EVT_USER_ASSOCIATION_REQUEST");
+		mosh_warn("Add the device to nRF Cloud and reconnect");
+		break;
+	case NRF_CLOUD_EVT_USER_ASSOCIATED:
+		mosh_print("NRF_CLOUD_EVT_USER_ASSOCIATED");
+		break;
+	default:
+		mosh_error("Unknown nRF Cloud event type: %d", evt->type);
+		break;
+	}
+}
+
+static void cmd_cloud_connect(const struct shell *shell, size_t argc, char **argv)
+{
+	int err;
+	static bool initialized;
+
+	if (!initialized) {
+		struct nrf_cloud_init_param config = {
+			.event_handler = nrf_cloud_event_handler,
+		};
+
+		err = nrf_cloud_init(&config);
+		if (err == -EACCES) {
+			mosh_print("nRF Cloud module already initialized");
+		} else if (err) {
+			mosh_error("nrf_cloud_init, error: %d", err);
+			return;
+		}
+
+		initialized = true;
+#if defined(CONFIG_NRF_CLOUD_PGPS)
+		k_work_init(&notify_pgps_work, notify_pgps);
+#endif
+		k_work_init_delayable(&cloud_reconnect_work, cloud_reconnect_work_fn);
+	}
+
+	k_work_reschedule(&cloud_reconnect_work, K_NO_WAIT);
```

Comment:
This was changed to a scheduled work. I don't know if calling the work function would work in practice but I'll leave it to you which way to go:
`cloud_reconnect_work_fn(NULL);`

Reply:
I think I've called work functions directly without issues as long as it doesn't need information from the k_work parameter. Anyway, it was more like a note that you could still call it directly but up to you which way you want to do it.

Reply:
I haven't called work functions directly either, although technically possible. When called directly, the function is executed in a different thread. There are a few consequences, like the thread stack size is likely different and in some cases it might be possible that the function would be executed in two different threads at the same time. Also, when searching the code for places where the work is used, a direct work function call may easily be missed. I think it would be better to be consistent and always execute works using the `k_work` functions.